### PR TITLE
User login, authentication, and authorization

### PR DESCRIPTION
**Pull request by kgilles**
_Disclaimer: This got away from me a bit as I changed my approach a couple times. I have to say auth management is one of my weak points, so don't be afraid to tell me if I've messed up._

What this does it creates separate API client classes for client-side and server-side. Which allows it to store the authorization on both ends. Having the app use the proper one where expected. Also creates a client-side `UserContext` to globally store user data. Plus adds a login page and a transient logout page.

That's the main part. Even though this changes a lot of files, most files have minor edits to work with the new API setup.

We don't utilize the NextJS frontend API as it could interfere with the expected API pathing for backend requests. Next doesn't allow us to rename the API path, meaning it's locked to `/api` which unfortunately is the same path backend expects.

### Screenshots

Desktop login page with unatuh nav bar
image

Mobile login page with unauth nav bar
image

Authed nav bar
image
image

Login error example
image

### CODE REVIEW

File: src/app/login/page.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import LoginForm from '@/components/auth/login-form';
+
+const Login = () => (
+  <div className="flex flex-col items-center p-24 md:p-56 w-full">
+    <div className="w-500">
+      <H1>Login</H1>
+      <div className="text-sm mt-12 mb-24">
+        <BodyText>{'Don’t have an account? '}</BodyText>
+        <Link href="/signup">
+          <LinkText>Sign up</LinkText>
```
Comment (slickplaid) [inline src/app/login/page.tsx:14]: ideally some level of i18n support is provided rather than hard coding text into the layouts

**Review by slickplaid — COMMENTED** [review]
There doesn't appear to be any testing or logging capabilities built-in to the code as written.

While some of the code uses the term `jwt` in the variable names, none of the code implements any of the considerations that a modern JWT client/refresh token strategy would use. Some of my comments might be incorrect if the intention was to use traditional session cookie style vs JWTs.

Overall it looks relatively sound, but if the intention is to utilize JWTs, a significant refactor might be required and supported by additional functionality on the backend.

File: src/app/login/page.tsx (hunk 1 above)
Comment (kgilles) [reply src/app/login/page.tsx:14]: Agreed. It's on the [project board](https://github.com/orgs/sublinks/projects/4) under New. I'll move it to Backlog.

File: src/components/auth/login-form.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      return;
+    }
+
+    try {
+      await SublinksApi.Instance().login(username, password);
+      saveMyUserFromSite();
+    } catch (e) {
+      setError('Login attempt failed. Please try again.');
+      setIsSubmitting(false);
+    }
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/components/auth/login-form.tsx:48]: Correct, it was set up like this as the user will be redirected after the login is successful. Happens in the `useEffect` further up in the file when `myUser` first becomes defined. I didn't want to allow the user to click the button more than once on successful attempt.

File: src/components/auth/login-form.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    const formData = new FormData(event.currentTarget);
+    const username = formData.get('username') as string;
+    const password = formData.get('password') as string;
+
+    if (!username || !password) {
+      setError('Please enter your username and password');
+      setIsSubmitting(false);
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/components/auth/login-form.tsx:40]: Good point, will do

[the remark this answers is not in the document]
Comment (kgilles) [reply src/components/auth/login-form.tsx:40]: image

File: src/components/auth/login-form.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  return (
+    <form onSubmit={handleLoginAttempt} className="flex flex-col">
+      <div className="flex flex-col gap-16">
+        <InputField type="text" label="Username" name="username" id="username" placeholder="Username" showBorderPlaceholder disabled={isSubmitting} />
+        <InputField type="password" label="Password" name="password" id="password" placeholder="Password" showBorderPlaceholder disabled={isSubmitting} />
+      </div>
+      <div className="h-32">
+        {error && <ErrorText className="text-sm">{error}</ErrorText>}
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/components/auth/login-form.tsx:58]: Smart, thanks!

File: src/components/user-nav/index.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  const ref = useClickOutside<HTMLUListElement>(() => open && setOpen(false));
+
+  const userAvatar = myUser?.local_user_view.person.avatar;
+
+  useEffect(() => {
+    document.documentElement.classList.remove('dark');
+    if (theme === 'dark') {
+      document.documentElement.classList.add('dark');
+    }
+  }, [theme]);
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/components/user-nav/index.tsx:34]: Might as well while we're in this file

File: src/context/user.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+// Placeholders for function to make sure TS always considers it defined
+export const UserContext = createContext<UserContextState>({
+  clearMyUser: () => {},
+  saveMyUserFromSite: () => {}
+});
+
+const UserProvider = ({ children }: { children: React.ReactNode }) => {
+  const [myUser, setMyUser] = useState<MyUserInfo>();
+
+  const saveMyUserFromSite = async () => {
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/context/user.tsx:25]: If things work correctly after the backend API is completed, there shouldn't be a need for mock data living inside the code. We have a dev setup which creates mock data through Docker. Do you think we need test data in both places?

File: src/utils/api-client/base.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+const isServerSide = () => typeof window === 'undefined';
+
+const getApiHost = () => {
+  const envUrl = isServerSide()
+    ? process.env.NEXT_PUBLIC_SUBLINKS_API_BASE_URL
+    : process.env.NEXT_PUBLIC_SUBLINKS_API_BASE_PUBLIC_URL;
+
+  return envUrl || 'localhost:8080';
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/utils/api-client/base.ts:19]: This fallback is just to have something in case an admin sets things up without configuring the env variables.
The URL should always come from `NEXT_PUBLIC_SUBLINKS_API_BASE_URL` or `NEXT_PUBLIC_SUBLINKS_API_BASE_PUBLIC_URL`. But worst case we fallback and hope for the best.

Perhaps we should not fallback at all, and force admins to set these variables.

File: src/utils/api-client/client.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    const cookieStore = {
+      get: () => Cookies.get(AUTH_COOKIE_NAME),
+      set: (value: string, options?: object) => Cookies.set(AUTH_COOKIE_NAME, value, options),
+      remove: () => Cookies.remove(AUTH_COOKIE_NAME)
+    } as CookieStore;
+
+    this.setAuthCookieStore(cookieStore);
+    this.setAuthHeader();
+  }
```
[the remark this answers is not in the document]
Comment (slickplaid) [reply src/utils/api-client/client.ts:18]: Let me see if I can come up with a couple contrived examples and this is all related as well to https://github.com/sublinks/sublinks-frontend/pull/79/files#r1496397310

**Async operations in constructors**

Consider
```js
class UserProfile {
  constructor(userId) {
    this.userId = userId;
    this.profile = null;
    this.loadProfile(); // Async operation inside constructor
  }

  async loadProfile() {
    this.profile = await fetchUserProfile(this.userId); // Assume this fetches user profile data
  }
}

// Trying to use the instance immediately after creation
const user = new UserProfile('user123');
console.log(user.profile); // This will likely log `null` because `loadProfile` has not completed yet
```
vs static async factory method
```js
class UserProfile {
  // only set up the object
  constructor(userId, profile) {
    this.userId = userId;
    this.profile = profile;
  }

  // do something
  static async load(userId) {
    const profile = await fetchUserProfile(userId);
    return new UserProfile(userId, profile);
  }
}

// Usage
UserProfile.load('user123').then(user => {
  console.log(user.profile); // This will correctly log the user profile after it's loaded
});
```

First way you lose control of the flow from the consumer's perspective where in the second example, you are explicit in action and retain all control of the flow. 

**Idempotency**

Idempotency in constructors ensures that creating multiple instances of a class does not lead to side effects that alter the global state or the state of other instances in unexpected ways. An idempotent constructor allows you to instantiate a class multiple times, with each instance being independent of the others and not affecting the application's overall state or behavior unexpectedly.

**Contrived Example**

Consider a scenario where you have a Logger class that registers itself to a global registry every time an instance is created. This is a contrived example.

```js
// Global registry for logger instances
const loggerRegistry = [];

class Logger {
  constructor(name) {
    this.name = name;
    // Non-idempotent behavior: Registering this instance in a global registry
    loggerRegistry.push(this);
  }

  log(message) {
    console.log(`[${this.name}]: ${message}`);
  }
}

// Creating multiple instances of Logger
const logger1 = new Logger("Logger1");
const logger2 = new Logger("Logger2");

console.log(loggerRegistry.length); // Outputs: 2
```
In this simple example, the constructor's side effect (registering the instance in a global registry) may seem harmless. However, in a more complex application, such behavior can lead to unintended consequences, such as:

- Memory Leaks: If instances are continuously created and registered but not properly deregistered or deleted, it could lead to memory leaks.
- Unexpected Behavior: Other parts of the application that interact with the loggerRegistry might exhibit unexpected behavior due to the growing number of registered instances.
- Testing Challenges: Automated tests may fail or behave inconsistently because the global state is modified every time a Logger instance is created, making it hard to isolate tests.

**Making the Constructor Idempotent**

To make the constructor idempotent, you should avoid side effects that alter the global state or rely on external state changes. If registration in a global registry is necessary, consider making it an explicit action that the consumer of the Logger class must invoke, rather than an automatic side effect of instantiation:

```js
class Logger {
  constructor(name) {
    this.name = name;
  }

  // Method to manually register the logger
  register() {
    loggerRegistry.push(this);
  }

  log(message) {
    console.log(`[${this.name}]: ${message}`);
  }
}

const logger1 = new Logger("Logger1");
// Registration is now an explicit action
logger1.register();
```

File: src/context/user.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (slickplaid) [reply src/context/user.tsx:25]: Thanks for bringing up the Docker setup for mock data - that's super handy for dev and integration tests! The idea behind suggesting dependency injection, is all about making our lives easier when we get down to the nitty-gritty of unit testing. 

Here's the gist:

- Quick Unit Tests: Using mocks directly in our code lets us run unit tests super fast and without the Docker overhead. It's all about testing small bits in isolation, especially when we're tweaking error handling or checking out those edge cases.
- Working in Parallel: It also means frontend and backend can move forward without stepping on each other's toes. We can mock out what we expect from the backend and keep rolling on the frontend, even if the backend isn't ready yet.
- Simplifying CI: Sometimes, CI environments get cranky with Docker, so having mocks can smooth things out and keep our tests running smoothly.

It's not about ditching Docker mocks but just giving us more flexibility and speed for certain types of tests. Plus, it's great for quick checks and new folks can see how things are supposed to work without diving deep into the backend setup.

[the remark this answers is not in the document]
Comment (kgilles) [reply src/context/user.tsx:25]: All great points!

I believe all mock data required by a test should live in the test itself. Rather than being defined in the code that will in the end be shipped to the end-user. This would be added when we do add tests. For which I now created [a task](https://github.com/sublinks/sublinks-frontend/issues/91) to track.

> We can mock out what we expect from the backend and keep rolling on the frontend, even if the backend isn't ready yet.

For now we have some test data living together with the API client for this very reason. Sublinks API doesn't have feature parity with Lemmy yet, which causes some friction when building out the frontend.

I believe until the backend catches up, we can add such test data on a feature-basis when we start building on something new. Rather than providing it by default for each API request right away. Might be too big of an effort, unless there's an easy way to pull in such test data that I'm not aware of.

File: src/utils/api-client/base.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
+import { SublinksClient } from 'sublinks-js-client';
+
+export const AUTH_COOKIE_NAME = 'auth';
+const AUTH_TTL_MS = 365 * 24 * 60 * 60 * 1000; // 365 days
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/utils/api-client/base.ts:4]: This will likely change as we collaborate with the backend to find an agreeable TTL. As we might want the cookie and the backend JWT to have the same lifespan.

Right now this simply mimics Lemmy.

File: src/utils/api-client/server.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import { cookies } from 'next/headers';
+import SublinksApiBase, { AUTH_COOKIE_NAME, CookieStore } from './base';
+
+class SublinksApiServerSide extends SublinksApiBase {
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/utils/api-client/server.ts:4]: Thanks for the heads up! This class will strictly run on the frontend client and server side. Cookies won't be stored or managed by the backend. At least it isn't right now and I haven't heard of any considerations to take that direction.

However, we do want to replace the Lemmy-style authentication with our own in the future. And if that affects this implementation, then I believe we would do best attacking it at that point in time. Rather than trying to mitigate for it now.

File: package.json [hunk 11]
```diff
@@ -29,6 +30,7 @@
   },
   "devDependencies": {
     "@stylistic/eslint-plugin-js": "^1.6.1",
+    "@types/js-cookie": "^3.0.6",
```
[the remark this answers is not in the document]
Comment (kgilles) [reply package.json:33]: As of now we have no need to extract information from the token. It's strictly used as an auth header passed along as-is.

File: src/context/user.tsx [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    const site = await SublinksApi.Instance().Client().getSite();
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/context/user.tsx:26]: As this is a context rather than a UI component I won't add a loading state to it. But I will add error handling.

File: src/utils/api-client/client.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kgilles) [reply src/utils/api-client/client.ts:18]: This is a beautiful comment. Thank you! I'll update the class constructor.

File: src/utils/api-client/server.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  public static Instance() {
+    const currentInstance = SublinksApiServerSide.instance ?? new SublinksApiServerSide();
+    const authCookie = currentInstance.authCookieStore?.get();
+
+    // If user has been logged out client-side we're also logging them out server-side
+    if (!authCookie && currentInstance.hasValidAuth) {
+      currentInstance.logout();
+    }
```
[the remark this answers is not in the document]
Comment (kgilles) [reply src/utils/api-client/server.ts:31]: We discussed this in Matrix but to note it here as well: The "JWT" implementation does, as you say, not use JWT practices. It's simply a JWT string passed around as the auth token.

It's being done this way as we're currently copying how Lemmy does it. Our first goal is feature-parity with Lemmy.

We have plans to implement our own, safer auth flow in the future after we've launched the first version of Sublinks. When that happens, so will this frontend update to support proper JWT(or other) practices.

> This is because the server-side check might not differentiate between a deliberate logout action and the mere absence of a cookie in a particular client session.

Great point, I will rework this.

**Comment by kgilles** [comment]
@slickplaid A lot of changes coming through. Thanks for your patience!
Hopefully I've answered all your questions, and addressed your feedback and concerns.

File: src/context/user.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+      if (site.my_user) {
+        setMyUser(site.my_user);
+      }
+    } catch (e) {
+      logger.error('Failed to get site for user context', e);
+    }
+  };
+
+  const clearMyUser = () => setMyUser(undefined);
```
Comment (slickplaid) [inline src/context/user.tsx:38]: This doesn't seem to log me out or I'm not running the `npm run dev:docker` instance correctly.

Interestingly, there might also be an issue with the profile dropdown or indication of being logged in or out as well. 

mobile shows that I am logged out or don't have a profile action submenu:
image

wider views shows a profile even after logout is clicked:
image

both are from the same page without any network requests between taking screenshots

Network trace (HAR): 
[localhost_Archive [24-03-03 12-02-39].zip](https://github.com/sublinks/sublinks-frontend/files/14474032/localhost_Archive.24-03-03.12-02-39.zip)

If you're not able to reproduce, I can take additional steps to try and clear my nextjs cache or some other thing that might be causing this due to user error rather than a bug. I can't seem to pin down exactly why so far.

The profile page doesn't appear to respect my auth token though so maybe the browser and state on the server side are out of sync.
image

Even navigating to `/login` shows the login page for a brief second and then brings me back to the root as if I'm logged in.

Short video:

https://github.com/sublinks/sublinks-frontend/assets/25877/f5526592-820d-4da1-9727-1209d112c300

File: src/utils/api-client/base.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (slickplaid) [reply src/utils/api-client/base.ts:21]: sensible defaults are always nice to have and i'm fine with this I think

been working in compiled languages primarily for a while and this kind of thing is a big no-no usually to better conform to the 12-factor app stuff with a hard coded string deep in the logic or control flow portion of the code but this is javascript 🤣

File: src/context/user.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (slickplaid) [reply src/context/user.tsx:26]: > I believe all mock data required by a test should live in the test itself. Rather than being defined in the code that will in the end be shipped to the end-user.

100% agreed hopefully my previous comment didn't allude that I thought otherwise. I think we're on the same page here to resolve this comment.

**Review by slickplaid — CHANGES_REQUESTED** [review]

**Review by slickplaid — APPROVED** [review]
I don't want to block this going out because I think it satisfies the immediate need for authentication, but there is a few usability concerns I have. 

1. Delay in registering that the user is logged in on page refresh. Even locally, this takes 2-3 seconds to RTT to the `/api/v3/site` endpoint to have the user object available and modify the ui to show as a logged in user. The site should know the user is logged in and any SSR happening should reflect that state.
firefox_xZJnsH61CH
2. The page should know, at time of initial SSR what the state of the index `/` page should be based on the initial request to a "cold" index page due to the cookie being present on that initial request. 
The post ordering, what images need to be fetched, and the issue above all are concerns about loading the page without full customization for the user logged in and should be known/presented to the client in that state through a server-side rendering condition ideally OR do not load _anything_ except skeleton placeholders until it can be known by the server. We had a brief discussion about this in an [unrelated thread on matrix](vector://vector/webapp/#/room/!kDzhpvKgppqRxkjCae:discuss.online/$EB_jPDke0W4TntCYsMVp5CI_kBsg18jFba-e4j_iM8s) (not sure if those urls that copy out of element work here)
3. There is going to be a lot more user customization that needs to be tracked and accounted for once profiles get developed and the above two issues likely will need to get addressed with that iteration and a better system for that user customization should be accounted for in future iterations.
4. Still strongly believe a full JWT infra should be present before v1 or shortly after as an enhancement, but as a cookie/opaque token auth this satisfies the basic needs.
